File: src/IsbInterop/DerivedRequisiteFactory.vb
Imports IsbInterop.Data
Imports IsbInterop.Data.Proxies
Imports IsbInterop.DataTypes.Enumerable
Imports IsbInterop.Utils

''' <summary>
''' Фабрика реквизитов, унаследованных от IRequisite.
''' </summary>
Friend NotInheritable Class DerivedRequisiteFactory

  ''' <summary>
  ''' Создает реквизит.
  ''' </summary>
  ''' <param name="rcwRequisite">RCW-объект реквизита.</param>
  ''' <param name="scope">Область видимости.</param>
  ''' <returns>Реквизит с заданным типом.</returns>
  Public Shared Function CreateRequisite(rcwRequisite As Object, scope As IScope) As IRequisite
    Dim type = TReqDataType.rdtUnknown

    Dim dataType = CInt(ComUtils.GetRcwProperty(rcwRequisite, "DataType"))

    If [Enum].IsDefined(GetType(TReqDataType), dataType) Then
      type = DirectCast(dataType, TReqDataType)
    End If

    Select Case type
      Case TReqDataType.rdtPick
        Return New PickRequisite(rcwRequisite, scope)
      Case TReqDataType.rdtDate, TReqDataType.rdtInteger, TReqDataType.rdtNumeric,
          TReqDataType.rdtReference, TReqDataType.rdtString, TReqDataType.rdtText
        Return New Requisite(rcwRequisite, scope)
      Case Else
        Return New Requisite(rcwRequisite, scope)
    End Select
  End Function

  ''' <summary>
  ''' Приватный конструктор.
  ''' </summary>
  Private Sub New()
  End Sub
End Class

File: src/IsbInterop/DerivedRequisiteFactory.cs
﻿using IsbInterop.Data;
using IsbInterop.Data.Proxies;
using IsbInterop.DataTypes.Enumerable;
using IsbInterop.Utils;
using System;
using IsbInterop.Data.Proxies.Requisites;
using IsbInterop.Data.Requisites;

namespace IsbInterop
{
  /// <summary>
  /// Фабрика реквизитов, унаследованных от IRequisite.
  /// </summary>
  internal static class DerivedRequisiteFactory
  {
    /// <summary>
    /// Создает реквизит.
    /// </summary>
    /// <param name="rcwRequisite">RCW-объект реквизита.</param>
    /// <param name="scope">Область видимости.</param>
    /// <returns>Реквизит с заданным типом.</returns>
    public static IRequisite CreateRequisite(object rcwRequisite, IScope scope)
    {
      var type = TReqDataType.rdtUnknown;

      int dataType = (int)ComUtils.GetRcwProperty(rcwRequisite, "DataType");

      if (Enum.IsDefined(typeof(TReqDataType), dataType))
        type = (TReqDataType)dataType;

      switch (type)
      {
        case TReqDataType.rdtPick:
          return new PickRequisite(rcwRequisite, scope);
        case TReqDataType.rdtDate:
        case TReqDataType.rdtInteger:
        case TReqDataType.rdtNumeric:
        case TReqDataType.rdtReference:
        case TReqDataType.rdtString:
        case TReqDataType.rdtText:
        default:
          return new Requisite(rcwRequisite, scope);
      }
    }
  }
}
